src/lib/util/pair.c

```diff
@@ -1232,6 +1232,60 @@ int fr_pair_prepend_by_da(TALLOC_CTX *ctx, fr_pair_t **out, fr_pair_list_t *list
 	return 0;
 }
 
+/** Alloc a new fr_pair_t, adding the parent attributes if required
+ *
+ * A child pair will be added to the first available matching parent
+ * found.
+ *
+ * @param[in] ctx	to allocate new #fr_pair_t in
+ * @param[out] out	Pair we allocated.  May be NULL if the caller doesn't
+ *			care about manipulating the fr_pair_t.
+ * @param[in] list	in which to insert the pair.
+ * @param[in] da	of the attribute to create.
+ * @return
+ *	- 0 on success.
+ *	- -1 on failure.
+ */
+int fr_pair_append_by_da_parent(TALLOC_CTX *ctx, fr_pair_t **out, fr_pair_list_t *list, fr_dict_attr_t const *da)
+{
+	fr_pair_t	*vp, *container = NULL;
+
+	/*
+	 *  Walk up the dictionary finding / adding parent attributes first
+	 */
+	if (!fr_dict_attr_is_top_level(da)) {
```

Comment:
Perhaps this could be made iterative, like ` fr_pair_find_by_da_nested()` using a da_stack?

That would likely be faster and use less memory.